alerting/src/main/kotlin/org/opensearch/alerting/MonitorRunner.kt

```diff
@@ -707,4 +721,112 @@ object MonitorRunner : JobRunner, CoroutineScope, AbstractLifecycleComponent() {
+                logger.info("Search hits for shard_$shard is: ${sh.hits.size}")
+
+                createFindings(sh)
+
+                logger.info("Updating monitor: ${monitor.id}")
+                lastRunContext[shard] = maxSeqNo.toString()
+                val updatedMonitor = monitor.copy(lastRunContext = lastRunContext)
+                // note: update has to called in serial for shards of a given index.
+                updateMonitor(client, xContentRegistry, settings, updatedMonitor)
+            } catch (e: Exception) {
+                logger.info("Failed to run for shard $shard", e)
+            }
+        }
+    }
+
+    private fun getShardsCount(index: String): Int {
+        val allShards: List<ShardRouting> = clusterService.state().routingTable().allShards(index)
+        return allShards.size
+    }
+
+    private fun createRunContext(index: String): HashMap<String, Any> {
+        val lastRunContext = HashMap<String, Any>()
+        lastRunContext["index"] = index
+        val count = getShardsCount(index)
+        lastRunContext["shards_count"] = count
+
+        for (i: Int in 0 until count) {
+            val shard = i.toString()
+            val maxSeqNo: Long = getMaxSeqNo(index, shard)
+            lastRunContext[shard] = maxSeqNo
+        }
+        return lastRunContext
+    }
+
+    /**
+     * Get the current max seq number of the shard. We find it by searching the last document
+     *  in the primary shard.
+     */
+    private fun getMaxSeqNo(index: String, shard: String): Long {
+        val request: SearchRequest = SearchRequest()
+            .indices(index)
+            .preference("_shards:$shard")
+            .source(
+                SearchSourceBuilder()
+                    .version(true)
+                    .sort("_seq_no", SortOrder.DESC)
+                    .seqNoAndPrimaryTerm(true)
+                    .query(QueryBuilders.matchAllQuery())
+                    .size(1)
+            )
+        val response: SearchResponse = client.search(request).actionGet()
+        if (response.status() !== RestStatus.OK) {
+            throw IOException("Failed to get max seq no for shard: $shard")
+        }
+        if (response.hits.hits.isEmpty())
+            return -1L
+
+        return response.hits.hits[0].seqNo
+    }
+
+    private fun searchShard(index: String, shard: String, prevSeqNo: Long?, maxSeqNo: Long, query: String): SearchHits {
+        val boolQueryBuilder = BoolQueryBuilder()
+        boolQueryBuilder.filter(QueryBuilders.rangeQuery("_seq_no").gt(prevSeqNo).lte(maxSeqNo))
+        boolQueryBuilder.must(QueryBuilders.queryStringQuery(query))
+
+        val request: SearchRequest = SearchRequest()
+            .indices(index)
+            .preference("_shards:$shard")
+            .source(
+                SearchSourceBuilder()
+                    .version(true)
+                    .query(boolQueryBuilder)
+                    .size(10000)
```

Comment:
Is this a placeholder, or do we actually want to set a size limit here since that can cause some data to be missed